transports/quic/src/connection.rs

```diff
@@ -21,407 +21,87 @@
-        Poll::Ready(Ok(substream))
-    }
-
-    fn poll_close(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
-        let mut inner = self.state.lock();
-        if inner.connection.is_drained() {
-            return Poll::Ready(Ok(()));
-        }
-
-        for substream in inner.substreams.keys().cloned().collect::<Vec<_>>() {
-            let _ = inner.connection.send_stream(substream).finish();
-        }
-
-        if inner.connection.streams().send_streams() == 0 && !inner.connection.is_closed() {
-            inner
-                .connection
-                .close(Instant::now(), From::from(0u32), Default::default())
-        }
-        drop(inner);
+        let this = self.get_mut();
 
-        loop {
-            match ready!(self.poll_event(cx)) {
-                Some(quinn_proto::Event::ConnectionLost { .. }) => return Poll::Ready(Ok(())),
-                None => return Poll::Ready(Err(Error::EndpointDriverCrashed)),
-                _ => {}
-            }
-        }
+        let (send, recv) = futures::ready!(this.outgoing.poll_unpin(cx))?;
+        let connection = this.connection.clone();
+        this.outgoing = Box::pin(async move { connection.open_bi().await });
+        let substream = Substream::new(send, recv);
+        Poll::Ready(Ok(substream))
     }
-}
 
-impl Drop for Connection {
-    fn drop(&mut self) {
-        let to_endpoint = ToEndpoint::ProcessConnectionEvent {
-            connection_id: self.connection_id,
-            event: quinn_proto::EndpointEvent::drained(),
-        };
-        self.endpoint_channel.send_on_drop(to_endpoint);
+    fn poll(
+        self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Result<StreamMuxerEvent, Self::Error>> {
+        // TODO: If connection migration is enabled (currently disabled) address
+        // change on the connection needs to be handled.
+        Poll::Pending
     }
-}
-
-/// Mutex-protected state of [`Connection`].
-#[derive(Debug)]
-pub struct State {
-    /// The QUIC inner state machine for this specific connection.
-    connection: quinn_proto::Connection,
-
-    /// State of all the substreams that the muxer reports as open.
-    pub substreams: HashMap<quinn_proto::StreamId, SubstreamState>,
-
-    /// Waker to wake if a new outbound substream is opened.
-    pub poll_outbound_waker: Option<Waker>,
-    /// Waker to wake if a new inbound substream was happened.
-    pub poll_inbound_waker: Option<Waker>,
-    /// Waker to wake if the connection should be polled again.
-    pub poll_connection_waker: Option<Waker>,
-}
 
-impl State {
-    fn unchecked_substream_state(&mut self, id: quinn_proto::StreamId) -> &mut SubstreamState {
-        self.substreams
-            .get_mut(&id)
-            .expect("Substream should be known.")
+    fn poll_close(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.connection.close(From::from(0u32), &[]);
```

Comment:
Do we have to wait for connection to be closed? https://docs.rs/quinn/latest/quinn/struct.Connection.html#method.closed

Reply:
The docs of `close` say that all pending operations will immediately fail and delivery of data on unfinished streams is not guaranteed.

Looking at the previous code, we did call `finish` on all open streams so I'd suggest we do the same thing here. The implementation of this function should be a graceful close of the connection.

Reply:
I am surprised there isn't a more straight-forward way to do this. Surely we are not the only ones that want to gracefully close connections?

Can we ask upstream to provide us with a function?

Reply:
> @kpp are you saying that we can not `.await` the `Future` returned by `self.connection.close(From::from(0u32), &[])`?

`close` itself does not return a `Future` but there is `closed` which @kpp mentioned further up. However, the problem is that calling `close` instantly terminates all in-progress `Stream`s.

To finish those, we will have to keep around all `SendStream`s and call `finish` on them before actually calling `close` here.

One way of doing this is by sharing them via `Arc<Mutex>` with the `Connection`. I don't really see another way.

Reply:
> > @kpp are you saying that we can not `.await` the `Future` returned by `self.connection.close(From::from(0u32), &[])`?
> 
> `close` itself does not return a `Future` but there is `closed` which @kpp mentioned further up. However, the problem is that calling `close` instantly terminates all in-progress `Stream`s.

Thanks. I missed this.

I am fine proceeding here. In other words I am fine merging this pull request without a proper graceful close as long as we track the outstanding work in a follow-up issue.

Reply:
@kpp what do you think of https://github.com/kpp/rust-libp2p/pull/30?